File: luaScripts/get_location_wifi_pattern.lua
local M = {}

local gu = require("luaScripts.general_utils")

local HOME = os.getenv("HOME")

local CSV_FILE = HOME .. "/.config/hypr/ignore/locations_wifi_patterns.csv"

local CURRENT_LOCATION_FILE = HOME .. "/.config/hypr/ignore/current_location.txt"

--------------------------------------------------
-- WiFi
--------------------------------------------------

local function get_wifi_networks()
	local networks = {}

	local handle = io.popen("nmcli -t -f SSID device wifi list 2>/dev/null")

	if not handle then
		return networks
	end

	for ssid in handle:lines() do
		if ssid ~= "" then
			table.insert(networks, ssid)
		end
	end

	handle:close()

	return networks
end

--------------------------------------------------
-- Location patterns
--------------------------------------------------

local function load_locations()
	local locations = {}

	local file = io.open(CSV_FILE, "r")

	if not file then
		return locations
	end

	-- Skip CSV header.
	file:read("*line")

	for line in file:lines() do
		local location, pattern = line:match("^([^,]+),([^,]+)$")

		if location and pattern then
			table.insert(locations, {
				location = location,
				pattern = pattern,
			})
		end
	end

	file:close()

	return locations
end

--------------------------------------------------
-- Location detection
--------------------------------------------------

function M.find_location()
	local wifi_networks = get_wifi_networks()
	local locations = load_locations()

	for _, location in ipairs(locations) do
		for _, wifi in ipairs(wifi_networks) do
			if wifi:find(location.pattern, 1, true) then
				return location.location
			end
		end
	end

	return "Unknown"
end

--------------------------------------------------
-- Main
--------------------------------------------------

function M.main()
	local location = M.find_location()

	if not gu.write_file(CURRENT_LOCATION_FILE, location) then
		return nil, "Could not write location file"
	end

	print(location)

	return location
end

return M

File: scripts/get_locations/get_location_wifi_pattern.py
#!/usr/bin/env python3

import csv
import os
import subprocess
from pathlib import Path

CSV_FILE = os.path.expandvars("$HOME/.config/hypr/ignore/locations_wifi_patterns.csv")


def get_wifi_networks():
    result = subprocess.run(
        ["nmcli", "-t", "-f", "SSID", "device", "wifi", "list"],
        capture_output=True,
        text=True,
        check=True,
    )

    return {line.strip() for line in result.stdout.splitlines() if line.strip()}


def load_locations():
    locations = []

    with open(CSV_FILE, newline="", encoding="utf-8") as file:
        reader = csv.DictReader(file)

        for row in reader:
            locations.append(
                {
                    "location": row["location"],
                    "pattern": row["pattern"],
                }
            )

    return locations


def find_location():
    wifi_networks = get_wifi_networks()
    print(wifi_networks)

    a = load_locations()
    print(f"a = {a}")

    for location in load_locations():
        pattern = location["pattern"]

        for wifi in wifi_networks:
            if pattern in wifi:
                return location["location"]

    return "Unknown"


if __name__ == "__main__":

    location_enum_name = find_location()
    CURRENT_LOCATION_FILE = Path.home() / ".config/hypr/ignore/current_location.txt"
    CURRENT_LOCATION_FILE.parent.mkdir(parents=True, exist_ok=True)
    CURRENT_LOCATION_FILE.write_text(location_enum_name)
    print(location_enum_name)
